{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# RNA-seq入门实战（二）：上游数据的比对计数——Hisat2+ featureCounts 与 Salmon\n",
    "\n",
    "> 本节概览：\n",
    "> 1. `hisat2` + `featureCounts`:\n",
    "> 获取hisat2索引文件，`hisat2`比对和`samtools`格式转化，`featureCounts`计数得到`counts`表达矩阵\n",
    "> 2. `Salmon`:\n",
    "> `salmon index` 用`cdna.fa`建立索引，`salmon quant`对`clean fastq`文件直接进行基因定量\n",
    "> 3. 获取`ensembl_id`或`transcript_id`转化的对应文件"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 一、hisat2 + featureCounts\n",
    "1. 获取hisat2比对索引文件\n",
    "index官网下载地址[`Download | HISAT2 (daehwankimlab.github.io)`](https://link.zhihu.com/?target=http%3A//daehwankimlab.github.io/hisat2/download/%23h-sapiens)，下载并解压所需的 `mm10` 或 `grcm38` 的`index`文件\n",
    "\n",
    "```shell\n",
    "mkdir -p ~/reference/index/hisat/\n",
    "cd ~/reference/index/hisat/\n",
    "wget https://genome-idx.s3.amazonaws.com/hisat/mm10_genome.tar.gz\n",
    "tar -zxvf *tar.gz \n",
    "rm *tar.gz\n",
    "## 或\n",
    "conda install hisat2\n",
    "conda install -c bioconda subread\n",
    "```\n",
    "\n",
    "2. `hisat2`比对和`samtools`转化格式\n",
    "\n",
    "**非常重要的步骤‼️**\n",
    "\n",
    "先用`hisat2`比对基因组得到sam文件，再用`samtools sort`将sam文件格式转化与排序为`bam`文件（bam相当于二进制版的sam），之后`samtools index`建立索引（用于后续IGV内可视化），最后`samtools flag` 统计文件比对情况保存在文本中。其中`samtools index`与`samtools flag`为非必须步骤，可略过。`sam`相当于是中间文件比较占存储空间，可在转化为bam后便删除。\n",
    "\n",
    "\n",
    "### hisat2 参数\n",
    "\n",
    "hisat2主要参数:\n",
    "- -x 参考基因组索引文件的前缀。注意最后是 **/genome**\n",
    "- -1 双端䇷序结果的第一个文件。若有多组数据,使用逗号将文件分隔， Reads的长度可以不一致。\n",
    "- -2 双觊䇣序结果的第二个文件。若有多组数据,使用逗号将文件分隔,并且文件顺序要和-1参数对应. Reads的长度可以不一致。\n",
    "- -S 指定输出的SAM文件。\n",
    "- -t 输出time\n",
    "- -p 线程使用设置\n",
    "- -U 单端数据文件.若有多组数据,使用逗号将文件分隔.可以和-1、-2参数同时使用. Reads的长度可以不一致。\n",
    "- -sra-acc 输入`SRA`登录号，比如SRR12207279，SRR12207280，若有多组数据,使用逗号将文件分隔，HISAT将启动下载并识别数据类型，进行对比\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## samtools\n",
    "```shell\n",
    "conda install samtools\n",
    "```\n",
    "\n",
    "### 参数\n",
    "#### 3.1 samtools sort 进行格式转化和排序\n",
    "\n",
    "```shell\n",
    "ls *.sam | while read id ;do samtools sort -@ 15 -o ~/wk/align/$(basename $id '.sam')_sort.bam $id ;\n",
    "done\n",
    "\n",
    "```\n",
    "- -@ INT 设置排序和压缩线程个数，默认是单线程\n",
    "- -o FILE 设置最终排序后的输出文件名\n",
    "- -I INT 设置输出文件压缩等级。0-9，0是不压缩，9是压缩等级最高。不设置此参数时，使用默认压缩等级；\n",
    "- -m INT 设置每个线程运行时的内存大小，可以使用K,M和G表示内存大小\n",
    "- -n 设置按照read名称进行排序\n",
    "- -O FORMAT 设置最终输出的文件格式，可以是bam,sam或者cram，默认为bam.\n",
    "- T PREFIX  设置临时文件前缀\n",
    "\n",
    "\n",
    "less命令可以正常查看sam文件，但是不能正常查看bam文文件，因为bam文文件是二进制文件，所以需要使用\n",
    "\n",
    "`samtools view *.bam|less -s`  查看bam文件     \n",
    "\n",
    "`samtools view SRR1039508_sort.bam chrX:24001837-24045303|less -SN`  查看bam文件特定区域"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### 3.2 samtools index 建立索引\n",
    "\n",
    "**步骤**：必须对`bam`文件进行默认情况下的排序后，才能进行index。否则会报错。建立索引后将产生后缀为`.bai`的文件，用于快速的随机处理。很多情况下需要有`bai`文件的存在，特别是显示序列比对情况下。比如`samtools`的`tview`命令就需要；`gbrowse2`显示 `reads`的比对图形的时候也需要。`IGV`显示比对情况也需要。\n",
    "\n",
    "```shell\n",
    "ls *.bam | xargs -i samtools index -@ 15 {}\n",
    "```\n",
    "- -b  创建一个BAI索引。当不使用格式选项时，这是当前的默认设置\n",
    "- -c  创建CSI索引。默认情况下 ，索引的最小间隔大小为2^14，与BAI格式使用的固定值相同\n",
    "- m INT # 创建CSI索引，最小格式大小2^INT\n",
    "- -@, --threads INT 除了主线程[0]外，要使用的输入/输出压缩线程数\n",
    "\n",
    "\n",
    "基因组可视化：`samtools tview  *.bam`  查看染色体区域的比对情况\n",
    "\n",
    "使用工具IGV  `integrative Genomics Viewer`\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 3.3 samtools flagstat 统计文件比对情况 并保存在文本中.flagstat\n",
    "\n",
    "```shell\n",
    "ls *.bam | while read id ; do samtools flagstat -@ 15 $id > $(basename $id '.bam').flagstat ; done\n",
    "```\n",
    "- -@ INT # 设置读取文件时要使用的额外线程数。\n",
    "- -O FORMAT # 设置输出格式。FORMAT可以设置为'default'， 'json'或'tsv'来选择默认的，json或标签分隔值输出格式，如果不使用此选项，将选择飘认格式。\n",
    "\n",
    "```shell\n",
    "ls *stat |while read id; do echo $id; sed \"s/5)/5) \\n/g\" $id; done\n",
    "# 展示`flagstat`文件，并显示出文件名，不同文件间空行（即把最后字符后添加到换行符）\n",
    "```\n",
    "\n",
    "`multiqc ./`  生成网页版的统计情况\n",
    "\n",
    "绘制成excel 表格展示统计情况:\n",
    "\n",
    "```shell\n",
    "cat *.flagstat |awk '{print $1}'|paste ------------   \n",
    "\n",
    "# 按照每行13个数字方式排列flagstat数据\n",
    "```\n",
    "\n",
    "\n",
    "`ctrl+c` 复制到excel表格，表格右下角选择 '使用文本导入向导' 将数据分割到小格中，复制粘贴进行转置\n",
    "\n",
    "\n",
    "```shell\n",
    "cat SRR1039508_sort.flagstat |awk -F '+ 0 ' '{print $2}'\n",
    "# 输出文本数据解释的部分\n",
    "# 粘贴至excel表中做成表格，保存为mytest文件\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### shell 脚本\n",
    "`vim  3_align2sam2bam_hisat2.sh`\n",
    "```shell\n",
    "\n",
    "############################\n",
    "echo -e  \" \\n \\n \\n 333#  Align !!! hisat2 !!!\\n \\n \\n \"\n",
    "date\n",
    "########set#### ###set#### ###set####   \n",
    "index='/home/reference/index/hisat/mm10/genome'\n",
    "\n",
    "mkdir  -p   /slurm/home/admin/nlp/DL/97-bioinformatics/gene_process_for_python/data/align/flag\n",
    "cd /slurm/home/admin/nlp/DL/97-bioinformatics/gene_process_for_python/data/align/\n",
    "pwd\n",
    "cat gene_process_for_python/data/idname | while read id\n",
    "do\n",
    "         echo \"333#  ${id} ${id} ${id}  is on the hisat2 Working !!!\"\n",
    "################ paired ###############################         \n",
    "            hisat2 -t -p 12 -x  $index \\\n",
    "       -1 /slurm/home/admin/nlp/DL/97-bioinformatics/gene_process_for_python/data/raw/clean/${id}_*1*gz  \\\n",
    "       -2  /slurm/home/admin/nlp/DL/97-bioinformatics/gene_process_for_python/data/raw/clean/${id}_*2*gz  -S  ${id}.sam\n",
    "######################################################\n",
    "################Single################################\n",
    "#              hisat2 -t -p 12 -x  $index  \\\n",
    "#                     -U /slurm/home/admin/nlp/DL/97-bioinformatics/gene_process_for_python/data/raw/clean/${id}_trimmed.fq.gz \\\n",
    "#                     -S  ./${id}.sam  \n",
    "########################################################           \n",
    "\n",
    "### sam2bam and remove sam ###\n",
    "    echo -e  \" ${id} sam2bam and remove sam   \"\n",
    "    samtools sort -@ 12 -o  /slurm/home/admin/nlp/DL/97-bioinformatics/gene_process_for_python/data/align/${id}_sorted.bam  ${id}.sam\n",
    "    rm ${id}.sam\n",
    "done\n",
    "\n",
    "#### samtools index and flagstat ####\n",
    "echo -e  \" \\n \\n \\n samtools index and flagstat \\n  \" \n",
    "cd   /slurm/home/admin/nlp/DL/97-bioinformatics/gene_process_for_python/data/align/flag\n",
    "pwd\n",
    "#ls /slurm/home/admin/nlp/DL/97-bioinformatics/gene_process_for_python/data/align/*.bam | xargs  -i  samtools index  -@  12  {}    \n",
    "ls /slurm/home/admin/nlp/DL/97-bioinformatics/gene_process_for_python/data/align/*.bam  | while read id ;\\\n",
    "do\n",
    "        samtools flagstat -@ 12 $id > $(basename $id '.bam').flagstat\n",
    "done\n",
    "multiqc ./\n",
    "\n",
    "echo -e \" \\n \\n \\n  333# ALL  Work Done !!!\\n \\n \\n \"\n",
    "date\n",
    "\n",
    "```\n",
    "\n",
    "`nohup bash 3_align2sam2bam_hisat2.sh >log_3 2>&1 &`"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "\n",
    "`grep alignment  log_3`       \n",
    "\n",
    "比对结果如下：\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 3. featureCounts 计数得到counts表达矩阵\n",
    "计数首先要获取gtf注释文件 **，注意要和hisat2的index文件的基因组版本相对应** ，如本次为`mm10`，则`gtf`文件也必须为`mm10`或`grcm38`。 研究人和鼠推荐用`gencode数据库`的文件`GENCODE` ，比较常用的还有UCSC的`refGene.gtf`文件，下载地址在https://hgdownload.soe.ucsc.edu/ （若想下载其他`gtf`文件则将网址中`mm10`替换即可，如`hg38`）。\n",
    "\n",
    "`featureCounts`详细使用方法见转录组定量可以用替换[`featureCounts`代替`HTSeq-count (qq.com)`](https://link.zhihu.com/?target=https%3A//mp.weixin.qq.com/s%3F__biz%3DMzAxMDkxODM1Ng%3D%3D%26mid%3D2247485726%26idx%3D1%26sn%3Deab040119aa041d5cdcd6e1c81179455%26scene%3D21%23wechat_redirect)，常用参数如下：\n",
    "\n",
    "### featureCounts的使用\n",
    "\n",
    "```shell\n",
    "Usage: featureCounts [options] -a <annotation_file> -o <output_file> input_file1 [input_file2]\n",
    "\n",
    "```\n",
    "生成counts文档，同时将命令运行结果输出结果日志（1为标准输出， 2 为失败输出）\n",
    "\n",
    "```shell\n",
    "featureCounts -T 15 -p -a ~ref/ucsc/hg19.refGene.gtf.gz -o counts.txt  ~/wk/align/*.bam  1> counts.log  2>&1\n",
    "\n",
    "multiqc ./ # 生成网页版统计情况\n",
    "```\n",
    "\n",
    "- -a 输入GTF/GFF基因组注释文件\n",
    "- -p 这个参数是针对paired-end数据\n",
    "- -F指定-a注释文件的格式，默认是GTF\n",
    "- -g 从注释文件中提取Meta-features信息用于read count, 默认是gene_id\n",
    "- -t 跟-g—样的意思，其是默认将exon作为一个feature\n",
    "- -o 输出文件\n",
    "- -T 多线程数\n",
    "\n",
    "\n",
    "`vim 4_counts_featurecounts.sh`\n",
    "\n",
    "\n",
    "```shell\n",
    "###########################################\n",
    "echo -e \"\\n \\n \\n444# Count #featureCounts !!! \\n \\n \\n\"\n",
    "date\n",
    "#####set####set###set\n",
    "gtf='/slurm/home/admin/nlp/DL/97-bioinformatics/gene_process_for_python/data/gencode.vM25.chr_patch_hapl_scaff.annotation.gtf'\n",
    "#gtf='/home/reference/gtf/UCSC/mm10.refGene.gtf.gz'\n",
    "\n",
    "mkdir  -p  /slurm/home/admin/nlp/DL/97-bioinformatics/gene_process_for_python/data/counts\n",
    "cd  /slurm/home/admin/nlp/DL/97-bioinformatics/gene_process_for_python/data/counts\n",
    "pwd\n",
    "######## single##########################################################################\n",
    "#featureCounts -T 12 -a $gtf  -o counts.txt  ~/test/align/*.bam        \n",
    "#######paired###########################################################################\n",
    "featureCounts -T  12  -p  -a  $gtf  -o  counts.txt  /slurm/home/admin/nlp/DL/97-bioinformatics/gene_process_for_python/data/align/*.bam\n",
    "#######################################################################################            \n",
    "###生成网页版统计情况\n",
    "multiqc ./\n",
    "\n",
    "echo -e \" \\n \\n \\n ALL WORK DONE !!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!  \\n \"\n",
    "date\n",
    "\n",
    "```\n",
    "\n",
    "\n",
    "`nohup bash 4_counts_featurecounts.sh >log_4 2>&1 &`"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "查看计数的网页版统计情况，匹配率在71-80%左右，还可以\n",
    "<image src=\"../image/19.png\">\n",
    "\n",
    "运行结果保存在counts文件夹下，里面的counts.txt就是我们下游分析所需要的文件啦"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 二、Salmon——直接对基因进行定量的工具"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "与hisat2不同，salmon不经过比对计数步骤而是直接对基因进行定量，如果不研究新转录本，用salmon方法可以更快更方便得到基因定量信息。\n",
    "\n",
    "```shell\n",
    "conda install salmon -c bioconda \n",
    "```\n",
    "\n",
    "\n",
    "salmon相比含有hisat2和STAR等软的RNA-seq流程而言，速庭更快，这是因为该软件基于转录组序列reference（也即是cDNA序\n",
    "列）并且基于k mert比对原理。\n",
    "通常如果是研究RNA-seq过程基因发生了何种变化且不需知道新转录本，可以直接使用salmon获取转录本的丰富信息。\n",
    "salmon是不基于比对计数而直接对基因进行定量的工具，适用于转录组、宏基因组等的分析。見原理摡括来讲是通过构建统计模型来推测已经注释的转录本呈现出什么表达模式时我们才会界序产生当前的FASTQ数据。其优势是：\n",
    "- 定量时考虑到不同样品中基因长度的改变 (比如不同isoform目的使用)\n",
    "- 速度快、需要的计算资涪和存㒂资源小\n",
    "- 敏感性高，不会丢弃目百到多个基因同源区域的reads\n",
    "- 可以直接校正GC-bias\n",
    "- 自动判断文库类型"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 1. 建立salmon索引\n",
    "先下载参考转录本序列cDNA.fa文件，在ensembl官网选择相应文件 [Index of /pub/release-102/fasta/mus_musculus/cdna/ (ensembl.org)](https://link.zhihu.com/?target=http%3A//ftp.ensembl.org/pub/release-102/fasta/mus_musculus/cdna/)，使用salmon index 建立索引文件，`salmon index`常用参数：\n",
    "\n",
    "- -p 线程\n",
    "- -t 接参考转录本序列\n",
    "- -i 设置建立索引的位置\n",
    "- -d 接decoy 件\n",
    "\n",
    "```shell\n",
    "mkdir ~/reference/index/salmon/grcm38\n",
    "cd ~/reference/index/salmon/grcm38\n",
    "salmon index -p 12 -t  ~/reference/ensembl/grcm38.cdna.fa.gz  -i  ./\n",
    "```\n",
    "\n",
    "\n",
    "### salmon定量基因\n",
    "使用`salmon quant`命令对`clean fastq`文件直接进行基因定量，主要参数如下：\n",
    "\n",
    "- -i 接之前生成的`salmon index`\n",
    "- -p 线程\n",
    "- -I [-libType] 描述库的格式字符串，-I A 设置软件自动判断库类型\n",
    "- -r [-unmatedReads] 包含末配对读取的文件列表 (如单头读取)\n",
    "- -1 [-mates1] 包含 #1 mates的文件q\n",
    "- -2 [-mates2] 包含 #2 mates的文件\n",
    "- -o           设置输出结果的文件名\n",
    "- `-g /--geneMap`  #提供转录本与基因之间对应关系的文件。\n",
    " #如果提供了该文件，`Salmon`将会输出文件：`quant.sf` 和 `quant.genes.sf`。            \n",
    " #其中quant.genes.sf 文件中含有对基因表达水平的估计。               \n",
    " #提供的文件可以是GTF文件，也可是由TAB分隔符分隔文件，文件中的每一行含转录本名称和对应的基因名。              \n",
    " #以'.gtf', '.gff' 或 '.gff3'结尾的文件名均被当做GTF格式做处理解析。              \n",
    " #以其他字符结尾的被当做由TAB分隔符分隔的简单文件格式做处理解析。           \n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "`vim 3333_salmon.sh`\n",
    "\n",
    "```shell\n",
    "###################################################\n",
    "echo -e  '\\n \\n \\n ### salmon quant is Working !!! \\n \\n'\n",
    "###set##set###set#############\n",
    "index=\"/home/reference/index/salmon/grcm38/\"\n",
    "\n",
    "mkdir /slurm/home/admin/nlp/DL/97-bioinformatics/gene_process_for_python/data/salmon\n",
    "cd /slurm/home/admin/nlp/DL/97-bioinformatics/gene_process_for_python/data/salmon\n",
    "pwd\n",
    "cat /slurm/home/admin/nlp/DL/97-bioinformatics/gene_process_for_python/data/idname | while read id \n",
    "do\n",
    "            echo \" '\\n  !!!!!! Processing sample $id !!!!! '\\n\" \n",
    "########single#############################################\n",
    "# salmon quant -i $index -l A \\\n",
    "#             -r ~/test/clean/$(basename $id)_trimmed.fq.gz \\\n",
    "#             -p 12  -o  $(basename $id)_quant\n",
    "#######paired#############################################\n",
    "salmon quant -i $index -l A \\\n",
    " -1 /slurm/home/admin/nlp/DL/97-bioinformatics/gene_process_for_python/data/clean/${id}_*1*gz \\\n",
    " -2 /slurm/home/admin/nlp/DL/97-bioinformatics/gene_process_for_python/data/clean/${id}_*2*gz  \\\n",
    " -p 12  --output ${id}_quant\n",
    "###############################################################     \n",
    "done\n",
    "\n",
    "multiqc ./\n",
    "\n",
    "echo -e \" \\n \\n \\n !!!!ALL WORK DONE !!!!!!!!!!!!!!!!!!!!! \\n\"\n",
    "date\n",
    "\n",
    "```\n",
    "\n",
    "`nohup bash 3333_salmon.sh  >log_333salmon  2>&1 &`\n",
    "\n",
    "运行结果存放在`salmon`文件夹下，里面的`quant.sf`即为下游分析所需要的文件"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 三、获取基因ID转化的对应文件\n",
    "\n",
    "由于本次使用的为gencode或ensembl的gtf与cdna文件，因此最后得到的为ensembl_id (gene_id)和 transcript_id，形式为：ENSMUSG00000000001.1 ，而我们下游常用gene symbol进行展示，因此还需要从gtf注释文件中获取ensembl_id 、transcript_id与gene symbol的对应关系文件。 方法如下：\n",
    "\n",
    "`vim gtf_geneid2symbol_gencode.sh`\n",
    "\n",
    "```shell\n",
    "#提取gtf注释文件中gene_id等与gene_name的对应关系,便于下游id转换\n",
    "gtf=\"gencode.vM25.chr_patch_hapl_scaff.annotation.gtf\"\n",
    "\n",
    "### gene_id to gene_name\n",
    "grep 'gene_id' $gtf | awk -F 'gene_id \\\"' '{print $2}' |awk -F '\\\"' '{print $1}' >gene_id_tmp\n",
    "grep 'gene_id' $gtf | awk -F 'gene_name \\\"' '{print $2}' |awk -F '\\\"' '{print $1}' >gene_name_tmp\n",
    "paste gene_id_tmp gene_name_tmp >last_tmp\n",
    "uniq last_tmp >g2s_vm25_gencode.txt\n",
    "rm *_tmp\n",
    "\n",
    "### transcript_id to gene_name\n",
    "grep 'transcript_id' $gtf | awk -F 'transcript_id \\\"' '{print $2}' |awk -F '\\\"' '{print $1}' >gene_id_tmp\n",
    "grep 'transcript_id' $gtf | awk -F 'gene_name \\\"' '{print $2}' |awk -F '\\\"' '{print $1}' >gene_name_tmp\n",
    "paste gene_id_tmp gene_name_tmp >last_tmp\n",
    "uniq last_tmp >t2s_vm25_gencode.txt\n",
    "rm *_tmp\n",
    "\n",
    "```\n",
    "\n",
    "`bash gtf_geneid2symbol_gencode.sh`\n",
    "\n",
    "所得文件如下所示\n",
    "\n",
    "<image src=\"../image/20.png\">"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "上游流程到此就结束了，将最后得到的`counts`文件夹与`g2s_vm25_gencode.txt` 或 `salmon`文件夹与`t2s_vm25_gencode.txt`下载到本地就可以愉快地进行下游分析了"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": []
  }
 ],
 "metadata": {
  "language_info": {
   "name": "python"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
